### ISSUE TITLE
Simultaneous calls send many requests

### ISSUE THREAD

**Issue by franciscop-invast** [auth0/lock#1314]
Multiple simultaneous and similar calls to `getUserInfo()` should not send several requests. At least not until the first one is resolved. To reproduce the bug, try this as a single statement after logging in:

```js
lock.getUserInfo(token, function(err, user) {});
lock.getUserInfo(token, function(err, user) {});
lock.getUserInfo(token, function(err, user) {});
lock.getUserInfo(token, function(err, user) {});
```

**Expected result**: a single request to `/authinfo`, all of them resolve correctly with the user data (or the error if there was any).

**Actual result**: 4 different calls to `/authinfo`.

**Proposed fix**: [something like this demo](https://jsfiddle.net/5rwq0zmz/9/) I made quickly:

```js
const pending = {};

const getUserInfo = async (token, cb) => {
  const key = '/userinfo' + token;
  if (pending[key]) return pending[key];
  pending[key] = fetch('/echo/json/').then(res => res.json()).then(data => {
    pending[key] = false;
    if (cb) cb(data);
    return data;
  });
  return pending[key];
};

getUserInfo('abcde');
getUserInfo('abcde');
getUserInfo('abcde');
getUserInfo('abcde');
getUserInfo('xxxxx');   // This will send a new one
```

**Workaround**: implement the above demo in userland.

It'd be much better if Auth0 implemented it though, it'd save Auth0 money (fewer repetitive requests) and make many sites that use Lock faster. A *cache* would be even better, but that'd be more of a Proposed Feature, while this issue is an unexpected behavior/bug.

**Lock version**: 11.4.0

**Comment by franciscop-invast** [comment]
As the workaround, put this into `async-lock.js`:

```js
import Lock from 'auth0-lock';

const cached = {};
const augment = ['getUserInfo'];

export default (...args) => {
  const plain = new Lock(...args);
  const lock = {};
  for (const key in plain) {
    lock[key] = !augment.includes(key) ? plain[key] : (...args) => {
      const cacheKey = key + args;
      if (cacheKey) return cached[cacheKey];
      cached[cacheKey] = new Promise((resolve, reject) => {
        plain[key](...args, (error, data) => {
          cached[cacheKey] = false;
          if (error) return reject(error);
          resolve(data);
        });
      });
      return cached[cacheKey];
    }
  }
  return lock;
};
```

All of the functions that you put in `augment` will be cached for the duration of the first request. In adition, they get promisified.

**Comment by luisrudge** [comment]
thanks for sharing your code snippet with others. that's up to the consumers of this library to handle specific scenarios like what you mentioned. I'm sure your code will help others when they face the same issue, thanks!

**Comment by franciscop-invast** [comment]
@luisrudge no problem! Just updated to fix a small bug. I think that:

- cache is normally client code, no question there.
- simultaneous requests as in my snippet is debatable, and it really depends on the library. Since `getUserInfo()` seems to be a thin wrapper for calling `/user/info`, then I agree that it is out of scope, but it would be nice to say in the docs that `getUserInfo()` is just a thin wrapper for an AJAX call.

However, could we get promises please? Want me to open a separate issue? It'd make working with lock much easier and simplify users code greatly with `async`/`await` by [avoiding callback hell](http://callbackhell.com/). It'd also make it trivial to add the functionality I mentioned above.

```js
// Default Lock lib:
const user = await lock.getUserInfo(token);
```

**Comment by luisrudge** [comment]
I'm a huge fan of promises, async/await, so you can bet that I'm interested in this as well. However, we're not adding promise support to auth0.js/lock in the near future.

**Comment by franciscop-invast** [comment]
@luisrudge internal policy? or resource constrain? (are you open to PR about that?)

**Comment by luisrudge** [comment]
it's a product decision. We don't want to add even more logic to both libraries, specially considering it would be conditional based on wether or not a callback is available (so we don't break anyone). Considering "promisifying" the library is super easy, we're not committed to have a promise-based API for now.
